File: Semestru 3/MAP/lab/lab2/src/model/SortingTask.java
package model;

import java.util.Vector;

public class SortingTask extends Task {
    private SortingMethod sortingMethod;
    private Vector<Integer> numbersVector;

    public SortingTask(String id_, String description_, SortingMethod sortingMethod_, Vector<Integer> numbersVector_) {
        super(id_, description_);
        sortingMethod = sortingMethod_;
        numbersVector = numbersVector_;
    }

    private static int partition(Vector<Integer> vector, int low, int high) {
        int pivot = vector.get(high);
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (vector.get(j).compareTo(pivot) > 0) {
                i++;
                int aux = vector.get(i);
                vector.set(i, vector.get(j));
                vector.set(j, aux);
            }
        }
        int aux = vector.get(i + 1);
        vector.set(i + 1, vector.get(high));
        vector.set(high, aux);
        return  i + 1;
    }

    private static void quickSort(Vector<Integer> vector, int low, int high) {
        if (low < high) {
            int pivot = partition(vector, low, high);
            quickSort(vector, low, pivot - 1);
            quickSort(vector, pivot + 1, high);
        }
    }

    private static void bubbleSort(Vector<Integer> vector) {
        boolean swap;
        for (int i = 0; i < vector.size() - 1; i++) {
            swap = false;
            for (int j = 0; j < vector.size() - 1 - i; j++)
                if (vector.get(j).compareTo(vector.get(j + 1)) < 0) {
                    var aux = vector.get(j);
                    vector.set(j, vector.get(j + 1));
                    vector.set(j + 1, aux);
                    swap = true;
                }
            if (!swap)
                break;
        }
    }

    @Override
    public void execute() {
        switch (this.sortingMethod) {
            case SortingMethod.BubbleSort -> {
                this.bubbleSort(this.numbersVector);
            }
            case SortingMethod.QuickSort -> {
                this.quickSort(this.numbersVector, 0, this.numbersVector.size() - 1);
            }
            default -> throw new IllegalStateException("Unexpected value: " + this.sortingMethod);
        }
    }

    public Vector<Integer> getNumbersVector() {
        return this.numbersVector;
    }

}

File: Semestru 3/MAP/lab/lab9/lab9/lab9/model/SortingTask.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace lab9.model
{
	internal class SortingTask : Task
	{
		private SortingMethod method;
		private List<int> numbersVector;

		public SortingTask(string id, string description, SortingMethod method_, List<int> numbersVector_)
			: base(id, description)
		{
			this.method = method_;
			this.numbersVector = numbersVector_;
		}

		private int partition(List<int> list, int low, int high)
		{
			int pivot = list[high];
			int i = low - 1;
			for (int j = low; j < high; j++)
			{
				if (list[j] > pivot)
				{
					i++;
					int temp = list[i];
					list[i] = list[j];
					list[j] = temp;
				}
			}
			int aux = list[i+1];
			list[i+1] =  list[high];
			list[high] = aux;
			return i + 1;
		}

		private void quickSort(List<int> list, int low, int high)
		{
			if (low < high)
			{
				int pivot = partition(list, low, high);
				quickSort(list, low, pivot - 1);
				quickSort(list, pivot + 1, high);
			}
		}

		private void bubbleSort(List<int> list)
		{
			bool swap;
			for (int i = 0; i < list.Count - 1; i++)
			{
				swap = false;
				for (int j = 0; j < list.Count - 1 - i; j++)
					if (list[j] < list[j + 1])
					{
						var aux = list[j];
						list[j] = list[j + 1];
						list[j + 1] = aux;
						swap = true;
					}
				if (!swap)
					break;
			}
		}

		public List<int> getNumbers() => this.numbersVector;

		public override void execute()
		{
			switch (this.method)
			{
				case SortingMethod.BubbleSort: 
					this.bubbleSort(this.numbersVector);
					break;
				case SortingMethod.QuickSort: 
					this.quickSort(this.numbersVector, 0, this.numbersVector.Count - 1);
					break;
				default: throw new ArgumentException("Unexpected value: " + this.method);
			}
		}
	}
}
